src/engine/strat_engine/mdv.rs

```diff
@@ -0,0 +1,142 @@
+use serde_json;
+use nix::unistd::fsync;
+
+use engine::engine::{HasUuid, FilesystemUuid};
+use engine::EngineResult;
+use engine::PoolUuid;
+use super::filesystem::{create_fs, mount_fs, unmount_fs, StratFilesystem};
+use super::serde_structs::{Isomorphism, FilesystemSave};
+
+// TODO: Monitor fs size and extend linear and fs if needed
+// TODO: Document format of stuff on MDV in SWDD (currently ad-hoc)
+
+#[derive(Debug)]
+pub struct MetadataVol {
+    dev: LinearDev,
+    mount_pt: PathBuf,
+}
+
+impl MetadataVol {
+    pub fn initialize(pool_uuid: &PoolUuid, dev: LinearDev) -> EngineResult<MetadataVol> {
+        try!(create_fs(dev.devnode().expect("must have a devnode").as_path()));
+        MetadataVol::setup(pool_uuid, dev)
+    }
+
+    pub fn setup(pool_uuid: &PoolUuid, dev: LinearDev) -> EngineResult<MetadataVol> {
+        let mount_pt = PathBuf::from(format!("/dev/stratis/.mdv-{}", pool_uuid.simple()));
+
+        // Mount point already exists? ok.
+        if let Err(err) = create_dir(&mount_pt) {
+            if err.kind() != ErrorKind::AlreadyExists {
+                return Err(From::from(err));
+            }
+        }
+
+        try!(mount_fs(&try!(dev.devnode()), &mount_pt));
+
+        // Subdirectory already exists? ok.
+        if let Err(err) = create_dir(&mount_pt.join("filesystems")) {
+            if err.kind() != ErrorKind::AlreadyExists {
+                return Err(From::from(err));
+            }
+        }
+
+        Ok(MetadataVol { dev, mount_pt })
+    }
+
+    /// Save info on a new filesystem to persistent storage, or update
+    /// the existing info on a filesystem.
+    // Write to a temp file and then rename to actual filename, to
+    // ensure file contents are not truncated if operation is
+    // interrupted.
+    pub fn save_fs(&self, fs: &StratFilesystem) -> EngineResult<()> {
+        let data = try!(serde_json::to_string(&fs.to_save()));
+        let path = self.mount_pt
+            .join("filesystems")
+            .join(fs.uuid().simple().to_string())
+            .with_extension("json");
+
+        let temp_path = path.clone().with_extension("temp");
+
+        // Braces to ensure f is closed before renaming
+        {
+            let mut f = try!(OpenOptions::new()
+                                 .write(true)
+                                 .create(true)
+                                 .open(&temp_path));
+            try!(f.write_all(data.as_bytes()));
+
+            // Try really hard to make sure it goes to disk
+            try!(f.flush());
+            try!(fsync(f.as_raw_fd()));
+        }
+
+        try!(rename(temp_path, path));
+
+        Ok(())
+    }
+
+    /// Remove info on a filesystem from persistent storage.
+    pub fn rm_fs(&self, fs_uuid: &FilesystemUuid) -> EngineResult<()> {
+        let fs_path = self.mount_pt
+            .join("filesystems")
+            .join(fs_uuid.simple().to_string())
+            .with_extension("json");
+        try!(remove_file(fs_path));
```

Comment:
I have a slight concern that this might return an error if filesystem is already gone, which would be undesirable. I can not tell immediately whether this is true or not from the implementation of ```remove_file``` or the docs.